---
title: "Chapter 6 Computational Activity"
output: html_notebook
---

#Chapter 6.3: Incomplete Lineage Sorting

###Preparation:
1. Set your working directory: 
```{r}

```

2. Load plotrix to your library.
```{r}
#install.packages("plotrix")
library(plotrix)
```

3. Reload the simulate.pop function into your R functions for use:
```{r}
#Reminders: const.RS refers to whether the reproductive success is constant for all individuals; mutation refers to whether mutations are allowed in the simulation or not; mut.rate is the mutation rate for an allele, for.class is a plotting setting. The initial state refers to the color setting of the alleles in the initial sample.

simulate.pop<-function(N.vec=rep(5,30), const.RS=TRUE,  mutation= TRUE, mut.rate=  0.1, for.class= TRUE, initial.state="all.black",plot.freqs=FALSE,mult.pop=FALSE,pops=FALSE){
   #  c(rep(10,5),rep(3,2),rep(10,5),rep(3,2),rep(10,5))  #
	stopifnot(initial.state %in% c("all.black","all.diff","two.alleles","single.mut") )

	if(plot.freqs){ layout(c(1,2)); par(mar=c(1,2,0,1))}
	if(for.class){
		line.lwd<-1
		line.col<-"black"
		mut.line.lwd<-1
		mut.line.col<-"black"
	
	}else{
		line.lwd<-0.5
		line.col<-"grey"
		mut.line.lwd<-1
		mut.line.col<-"grey"
	}

	num.gens<- length(N.vec)-1	

	if(!mult.pop){
		ind.pop.par<-matrix(1,nrow=max(N.vec),ncol=num.gens+1)
		ind.pop<-matrix(1,nrow=max(N.vec),ncol=num.gens+1)
	}else{
			ind.pop.par<-pops[["ind.pop.par"]]
			ind.pop<-pops[["ind.pop"]]
	}
	
	num.gens<- length(N.vec)-1
	offset<-0.1
	plot(c(1,num.gens),c(0.5,max(N.vec))+c(-offset,offset),type="n",axes=FALSE,xlab="",ylab="")
	mtext(side=1,line=0,"Generations")
	 text(1,0.5,"Past")
	 text(num.gens-1,0.5,"Present")
	 
	 track.cols<- list()
	N <-N.vec[1]
	 if(initial.state=="all.black") my.cols<-rep("black",2*N)  #sample(rainbow(2*N))
	 if(initial.state=="all.diff") my.cols<-sample(rainbow(2*N))
	 if(initial.state=="two.alleles")  my.cols<-  rep(c("blue","red"),N)
	 if(initial.state=="single.mut")  my.cols<-  c("red",rep("blue",2*N-1))
	 stopifnot((2*N)==length(my.cols))
	
	 track.cols[[1]]<-my.cols
	points(rep(1,N),1:N+offset, pch=19,cex=1.3,col=my.cols[(1:N)*2])
	 points(rep(1,N),1:N-offset, pch=19,cex=1.3,col=my.cols[(1:N)*2-1])
	
	for(i in 1:num.gens){
			
		N.new<-N.vec[i+1]
		N.old<-N.vec[i]
		points(rep(i,N.old),1:N.old+offset, pch=19,cex=1.3,col=my.cols[(1:N.old)*2])
		points(rep(i,N.old),1:N.old-offset, pch=19,cex=1.3,col=my.cols[(1:N.old)*2-1])

		 new.cols<-rep("black",2*N.new)
		
		if(const.RS){ 
			repro.success<-rep(1/N.old,N.old)
		}else{
			repro.success<-sample(c(rep(0.5/(N.old),N.old-2),c(0.25,0.25)),replace=FALSE)
			}
	
		 for(ind in 1:N.new){

				this.pop.par <- ind.pop.par[ind,i+1]
				available.pars <- (1:N.old)[which(ind.pop[1:N.old,i] == this.pop.par)]
				par<-sample(available.pars,2,replace=FALSE,prob=repro.success[which(ind.pop[1:N.old,i] == this.pop.par)])
		
				which.allele.1<-sample(c(-1,1),1)
				if(i != num.gens){ lines(c(i,i+1), c(par[1]+which.allele.1*offset,ind-offset),col=line.col,lwd=line.lwd)}
				new.cols[2*ind-1]<- my.cols[2*par[1] +ifelse(which.allele.1==1,0,-1)]
		
				which.allele.2<-sample(c(-1,1),1)
				if(i != num.gens){ lines(c(i,i+1), c(par[2]+which.allele.2*offset,ind+offset),col=line.col,lwd=line.lwd)}
				new.cols[2*ind]<- my.cols[2*par[2] +ifelse(which.allele.2==1,0,-1)]
		
				if(mutation){
					if(runif(1)<mut.rate){ 
							new.cols[2*ind-1]<- sample(rainbow(4*N),1)
							if(i != num.gens){ lines(c(i,i+1), c(par[1]+which.allele.1*offset,ind-offset),col=mut.line.col,lwd=mut.line.lwd)}
		
						}
					if(runif(1)<mut.rate){ 
						new.cols[2*ind]<- sample(rainbow(4*N),1)
						if(i != num.gens){ lines(c(i,i+1), c(par[2]+which.allele.2*offset,ind+offset),col=mut.line.col,lwd=mut.line.lwd)}
				} 
				
			}
		}	
		##redraw points to cover lines		 
		points(rep(i,N.old),1:N.old+offset, pch=19,cex=1.3,col=my.cols[(1:N.old)*2])
		points(rep(i,N.old),1:N.old-offset, pch=19,cex=1.3,col=my.cols[(1:N.old)*2-1])

		 my.cols<-new.cols
		 track.cols[[i+1]]<-my.cols
		if(!const.RS) sapply(which(repro.success>1/N.old), function(ind){ draw.circle(x=i,y=ind,radius=0.2,nv=100,border=NULL,col=NA,lty=1,lwd=1)})
	}
#	recover()
	if(plot.freqs){
		plot(c(1,num.gens),c(0,1),type="n",axes=FALSE,xlab="",ylab="")
		all.my.cols<-unique(unlist(track.cols))
		
		if(!mult.pop){ 
			my.col.freqs<-sapply(track.cols,function(my.gen){sapply(all.my.cols,function(my.col){sum(my.gen==my.col)})})

			sapply(all.my.cols,function(col.name){lines(my.col.freqs[col.name,]/(2*N.vec),col=col.name,lwd=2)});
			}else{
			
			for(pop in 1:max(ind.pop)){
				my.col.freqs<-sapply(1:num.gens, function(gen){
	#			recover()
					my.gen<-track.cols[[gen]]
					if(all(ind.pop.par[ind.pop[,gen]==pop,gen]==0)) return(rep(NA,length(all.my.cols)))  #if pop doesn't exist in this gen.
				
					these.inds<-which(ind.pop[,gen]==pop)
					my.gen<-c(my.gen[these.inds*2],my.gen[these.inds*2-1])
					sapply(all.my.cols,function(my.col){
						sum(my.gen==my.col)
					})})
				rownames(my.col.freqs)<-		all.my.cols
				sapply(all.my.cols[-length(all.my.cols)],function(col.name){lines(my.col.freqs[col.name,]/(2*5),col=col.name,lwd=2,lty=pop)});	
			}
			}
		
		axis(2)
	}
}
```

#Incomplete Lineage Sorting: Recreating Figures 6.4 + 6.5. 

##We will now incorporate multiple populations into the simulation with mult.pop=TRUE. 
```{r}
####Incomplete lineage sorting set-up
	one.out.of.3<-c(rep(1,5),rep(0,10)) #these parameters are setting up the stairstepping population splits in the first 15 generations of the simulation. This first one represents 1 population for 5 generations, then a split.
	two.out.of.3<-c(rep(1,5),rep(2,5),rep(0,5)) #this represents 1 population for 5 generations, then 2 populations for five generations
	three.out.of.3<-c(rep(1,5),rep(2,5),rep(3,5)) #this is the final part, which sets up the 3 populations for 5 generations following the final split.
	pops<-list()
	pops[["ind.pop.par"]]<-	matrix(c(rep(one.out.of.3,5),c(rep(1,10),rep(0,5)),rep(two.out.of.3,4),c(rep(1,5),rep(2,10)) ,rep(three.out.of.3,15)),nrow=15)	#this is setting up the population parameter. It creates a matrix representing the individuals and generations as they split from one to two and finally three populations.
	
	N.vec<-apply(pops[["ind.pop.par"]],2,function(x){sum(x!=0)}) #sets up final number of individuals and generations
	#N.vec<-apply(ind.pop.par,2,function(x){sum(x!=0)})
	num.gens<- length(N.vec)-1
    pops[["ind.pop"]]<-sapply(1:num.gens,function(i){c(rep(1,5),rep(2,5),rep(3,5))})
```

#Actual code to run simulation:
**Note: you will need to run this many times, I have set to 50 replicates)**
```{r}
	replicate(50,{ #set this number to the number of replicates you want to run.
		simulate.pop(N.vec=N.vec, const.RS=TRUE,  mutation= FALSE, for.class= TRUE, initial.state="single.mut",mult.pop=TRUE,pops=pops)
		lines(x=c(6.5,100),y=c(5.5,5.5),lwd=4,col="darkgrey")	##show barrier to migration
		lines(x=c(11.5,100),y=c(10.5,10.5),lwd=4,col="darkgrey")
	})
```

#What if we added another population? This builds our null expectation for incomplete lineage sorting mutational types: ABBA or BABA
```{r}
####Incomplete lineage sorting ABBA/BABA set-up
	one.out.of.4<-c(rep(1,5),rep(0,15)) 
	two.out.of.4<-c(rep(1,5),rep(2,5),rep(0,10)) 
	three.out.of.4<-c(rep(1,5),rep(2,5),rep(3,5),rep(0,5)) 
	four.out.of.4<-c(rep(1,5),rep(2,5),rep(3,5),rep(4,5)) 
	
	pops2<-list()
	pops2[["ind.pop.par"]]<-	matrix(c(rep(one.out.of.4,5),c(rep(1,10),rep(0,10)), c(rep(two.out.of.4,4),c(rep(1,5),rep(2,10),rep(0,5)),rep(three.out.of.4,4),c(rep(1,5),rep(2,5),rep(3,10)) ,rep(four.out.of.4,15))),nrow=20)	#done!
	
	N.vec2<-apply(pops2[["ind.pop.par"]],2,function(x){sum(x!=0)}) #sets up final number of individuals and generations
	#N.vec<-apply(ind.pop.par,2,function(x){sum(x!=0)})
	num.gens2<- length(N.vec2)-1
    pops2[["ind.pop"]]<-sapply(1:num.gens2,function(i){c(rep(1,5),rep(2,5),rep(3,5),rep(4,5))})
```

```{r}
	replicate(100,{ #set this number to the number of replicates you want to run.
		simulate.pop(N.vec=N.vec2, const.RS=TRUE,  mutation= FALSE, for.class= TRUE, initial.state="single.mut",mult.pop=TRUE,pops=pops2)
		lines(x=c(6.5,100),y=c(5.5,5.5),lwd=4,col="darkgrey")	##show barrier to migration
		lines(x=c(11.5,100),y=c(10.5,10.5),lwd=4,col="darkgrey")
		lines(x=c(16.5,100),y=c(15.5,15.5),lwd=4,col="darkgrey")
	})
```
#Given no gene flow, ABBA and BABA patterns should be equally likely to occur due to incomplete lineage sorting alone.

#However, if gene flow is occuring among two of any of the four lineages, we expect to see that the ratio of ABBA to BABA mutational patterns will change.

Specifically:
If we see more ABBA than BABA, we expect that gene flow has recently occurred between populations B and C (see above to look at simulations of what that would look like). This would result in the ratio from equation 6.11 being positive, as there are more ABBA than BABA patterns. This ratio is often termed Patterson's D, or the D-statistic.

If we see more BABA than ABBA, we expect that gene flow has recently occurred between populations A and C (see above to look at simulations of what that would look like). This would result in equation 6.11 being negative, as there are more BABA than ABBA patterns. 

#Now let's explore calculating the D-statistic using R with data included in the evobiR package.
```{r}
install.packages("evobiR")
library(evobiR)
```

#The CalcD command calculates the D-statistic of a sequence alignment. In this case, we call some example data ("1.fasta" from the evobiR package) and calculate the D-statistic without a significance test.
```{r}
CalcD(alignment = system.file("1.fasta", package = "evobiR"), sig.test = "N")
```
#Based on the above value, we can say that there has been gene flow between populations A and C of this dataset, as the statistic is negative, and thus there should be an abundance of the BABA mutational pattern. 

#Remind yourself, what mutational pattern is common when you see a positive D-statistic? What about a negative D-statistic? And finally, what would you expect D to be if there is no gene flow between populations after they split from each other?
